pkg/parser/node/matrix.go

```diff
@@ -0,0 +1,106 @@
+package node
+
+type Predicate func(nodeName string) bool
+
+func (node *Node) FindParent(predicate Predicate) *Node {
+	current := node.Parent
+
+	for current != nil {
+		if predicate(current.Name) {
+			return current
+		}
+
+		current = current.Parent
+	}
+
+	return nil
+}
+
+func ReversePath(path []int) []int {
+	var result []int
+
+	if len(path) == 0 {
+		return []int{}
+	}
+
+	for i := len(path) - 1; i >= 0; i-- {
+		result = append(result, path[i])
+	}
+
+	return result
+}
+
+func (node *Node) PathUpwardsUpto(upto *Node) []int {
+	var result []int
+
+	parent := node.Parent
+	lookingFor := node
+
+	for parent != nil {
+		for i, parentChild := range parent.Children {
+			if parentChild == lookingFor {
+				result = append(result, i)
+				lookingFor = parent
+				break
+			}
+		}
+
+		if parent == upto {
+			break
+		}
+
+		parent = parent.Parent
+	}
+
+	return result
+}
+
+func (node *Node) GetPath(path []int) *Node {
```

Comment:
Maybe make a type alias for `TreePath`s instead of using `[]int`?

Reply:
I don't think it'd make any difference, because virtually all path operations need to know that this is a slice of ints and on top of that all of this is only used internally.